File: module-infra/src/main/java/com/hanghae/cache/CacheManager.java
package com.hanghae.cache;

import java.time.Duration;
import java.time.LocalDateTime;
import java.util.Optional;

public interface CacheManager<K, V> {
    void put(K key, V value);
    void put(K key, V value, Duration expireAfter);
    Optional<V> get(K key);
    void remove(K key);
}

File: infrastructure/src/main/kotlin/com/hanghe/redis/cache/CacheManager.kt
package com.hanghe.redis.cache

import java.time.Duration
import kotlin.reflect.KClass

interface CacheManager {

    fun <T : Any> getOrNull(key: String, clazz: KClass<T>): T?

    fun <T : Any> getOrPut(key: String, ttl: Duration, clazz: KClass<T>, cacheable: () -> T): T

    fun <T : Any> put(key: String, ttl: Duration, clazz: KClass<T>, cacheable: () -> T): T
}
